compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -397,7 +397,11 @@ fn impl_intersection_has_negative_obligation(
 ) -> bool {
     debug!("negative_impl(impl1_def_id={:?}, impl2_def_id={:?})", impl1_def_id, impl2_def_id);
 
-    let ref infcx = tcx.infer_ctxt().intercrate(true).with_next_trait_solver(true).build();
+    // N.B. We don't need to use intercrate mode here because we're trying to prove
+    // the *existence* of a negative goal, not the non-existence of a positive goal.
+    // Without this, we over-eagerly register coherence ambiguity candidates when
+    // impl candidates do exist.
+    let ref infcx = tcx.infer_ctxt().with_next_trait_solver(true).build();
```

Comment:
I believe the intercrate mode must be enabled when equating impl headers in order to not over constrain the common type (e.g. disallow equating alias substs). It can then, and only then, be safely disabled in `try_prove_negated_where_clause`.

I think this is only a theoretical unsoundness for now because we currently ignore the obligations of `equate_impl_headers` and because `infcx.eq` does not have intercrate-specfic behavior (in the new solver mode that is. for the old solver, there is [this](https://github.com/rust-lang/rust/blob/ea6b131132eb383e8f4ae136fc39c4dec53b8ec6/compiler/rustc_infer/src/infer/combine.rs#L130)).

Reply:
Yeah, that makes sense. I can leave a comment clarifying that next to the line where we currently throw away these obligations.

Reply:
can't we just disable it in `try_prove_negated_where_clause`? the fact that we rely on `infcx.eq` to not have intercrate-specific behavior is a bit scary.

Reply:
I don't think we have facilities to set the intercrate mode currently, but I guess I could add it to `fork`, probably the best way to do it.